drivers/kw41zrf/kw41zrf_netdev.c

```diff
@@ -0,0 +1,837 @@
+{
+    /* get size of the received packet */
+    uint8_t pkt_len = (ZLL->IRQSTS & ZLL_IRQSTS_RX_FRAME_LENGTH_MASK) >> ZLL_IRQSTS_RX_FRAME_LENGTH_SHIFT;
+    kw41zrf_t *dev = (kw41zrf_t *)netdev;
+    if (pkt_len < IEEE802154_FCS_LEN) {
+        __asm__ volatile ("BKPT #1\n");
+//         return -EAGAIN;
+    }
+    /* skip FCS */
+    pkt_len -= IEEE802154_FCS_LEN;
+    DEBUG("[kw41zrf] RX %u bytes\n", pkt_len);
+
+    /* just return length when buf == NULL */
+    if (buf == NULL) {
+        if (len > 0) {
+            /* discard what we have stored in the buffer, go back to RX mode */
+            kw41zrf_set_sequence(dev, dev->idle_state);
+        }
+        return pkt_len;
+    }
+
+//     DEBUG("[kw41zrf] buf len: %3u\n", (unsigned int)pkt_len);
+#if defined(MODULE_OD) && ENABLE_DEBUG
+    DEBUG("[kw41zrf] recv:\n");
+    od_hex_dump((const uint8_t *)ZLL->PKT_BUFFER_RX, pkt_len, OD_WIDTH_DEFAULT);
+#endif
+
+#ifdef MODULE_NETSTATS_L2
+    netdev->stats.rx_count++;
+    netdev->stats.rx_bytes += pkt_len;
+#else
+    (void)netdev;
+#endif
+
+    if (pkt_len > len) {
+        /* not enough space in buf */
+        return -ENOBUFS;
+    }
+    memcpy(buf, (const void *)&ZLL->PKT_BUFFER_RX[0], pkt_len);
+
+    if (info != NULL) {
+        netdev_ieee802154_rx_info_t *radio_info = info;
+        uint8_t hw_lqi = (ZLL->LQI_AND_RSSI & ZLL_LQI_AND_RSSI_LQI_VALUE_MASK) >>
+            ZLL_LQI_AND_RSSI_LQI_VALUE_SHIFT;
+        /* TODO Validate, verify or adjust this LQI calculation */
+        if (hw_lqi >= 220) {
+            radio_info->lqi = 255;
+        } else {
+            radio_info->lqi = (51 * hw_lqi) / 44;
+        }
+        radio_info->rssi = (ZLL->LQI_AND_RSSI & ZLL_LQI_AND_RSSI_RSSI_MASK) >> ZLL_LQI_AND_RSSI_RSSI_SHIFT;
+    }
+
+    /* Go back to RX mode */
+    kw41zrf_set_sequence(dev, dev->idle_state);
+
+    return pkt_len;
+}
+
+static int kw41zrf_netdev_set_state(kw41zrf_t *dev, netopt_state_t state)
+{
+    switch (state) {
+        case NETOPT_STATE_OFF:
+        case NETOPT_STATE_SLEEP:
+            kw41zrf_set_power_mode(dev, KW41ZRF_POWER_DSM);
+            break;
+        case NETOPT_STATE_IDLE:
+            kw41zrf_set_power_mode(dev, KW41ZRF_POWER_IDLE);
+            kw41zrf_set_sequence(dev, dev->idle_state);
+            break;
+        case NETOPT_STATE_TX:
+            if (dev->netdev.flags & KW41ZRF_OPT_PRELOADING) {
+                kw41zrf_tx_exec(dev);
+            }
+            break;
+        case NETOPT_STATE_RESET:
+            kw41zrf_reset_phy(dev);
+            break;
+        default:
+            return -ENOTSUP;
+    }
+    return sizeof(netopt_state_t);
+}
+
+int kw41zrf_netdev_get(netdev_t *netdev, netopt_t opt, void *value, size_t len)
+{
+    kw41zrf_t *dev = (kw41zrf_t *)netdev;
+
+    if (dev == NULL) {
+        return -ENODEV;
+    }
+
+    switch (opt) {
+        case NETOPT_MAX_PACKET_SIZE:
+            if (len < sizeof(int16_t)) {
+                return -EOVERFLOW;
+            }
```

Comment:
As a suggestion, using `assert`s for these checks saved some code size in the ar86rf2xx driver.